### PR TITLE
Inject OpenShift custom ca bundle into console pod

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

- Bugfix

### Description

Inject custom CA bundle into Console pod so that Console authentication functions even when the OpenShift cluster uses a custom CA.

### Checklist

- [ ] Update/write design documentation in `./documentation/design`
- [x] Write tests and make sure they pass
- [ ] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

### CODE REVIEW

**Comment by lulf** [comment]
Should we support a way of explicitly configuring the console CA in the CRD as well? It could be useful both on Kubernetes and as an alternative fallback.

**Comment by k-wall** [comment]
> Should we support a way of explicitly configuring the console CA in the CRD as well? It could be useful both on Kubernetes and as an alternative fallback.

Yes. I think so.  I'm planning to add providerTrustBundle that references a configmap.  OpenShift will default its name and add the annotation (enmasse.io/ca-bundle-hash).  On Kubernetes, the user can provide the configmap if he wishes. In the configmap, key(s that match pattern *.crt will be assumed to contain a PEM(s).

The configmap will be optionally mounted by console-init.  It will aggregate all the trust into a single file and write to a location known to oauth-proxy.  oauth-proxy will read use this file for its provider trust.   There's a slight wrinkle with oauth-proxy - the OpenShift version requires a file, whereas the other one requires a cert directory.

File: systemtests/src/main/java/io/enmasse/systemtest/platform/Kubernetes.java [hunk 1]
```diff
@@ -532,6 +534,10 @@ public Pod getPod(String name) {
         return client.persistentVolumeClaims().inNamespace(infraNamespace).withLabels(labels).list().getItems();
     }
 
+    public List<Route> listRoutes(String namespace, Map<String, String> labels) {
```
Comment (lulf) [inline systemtests/src/main/java/io/enmasse/systemtest/platform/Kubernetes.java:537]: This should be in OpenShift.java I think and the default implementation should be throwing an exception maybe? (CC @kornys )

Comment (kornys) [reply systemtests/src/main/java/io/enmasse/systemtest/platform/Kubernetes.java:537]: @lulf yup I agree

Comment (k-wall) [reply systemtests/src/main/java/io/enmasse/systemtest/platform/Kubernetes.java]: Fixed.

**Comment by k-wall** [comment]
@enmasse-ci run tests testcase=FirefoxConsoleTest crc

File: systemtests/src/test/java/io/enmasse/systemtest/bases/web/ConsoleTest.java [hunk 2]
```diff
@@ -1419,6 +1424,134 @@ protected void doTestErrorDialog(AddressSpace addressSpace) throws Exception {
...
+        assertThat(console.isPresent(), is(true));
+
+        CertPair originalOauthCert = OpenSSLUtil.downloadCert(oauthRoute.get().getSpec().getHost(), 443);
+
+        String customCaCnName = "custom-ca";
+        String customIngressSecretName = "custom-ingress";
+        try (CertPair ca = OpenSSLUtil.createSelfSignedCert("/O=io.enmasse/CN=MyCA");
+             CertPair unsignedCluster = OpenSSLUtil.createSelfSignedCert("/O=io.enmasse//CN=MyCluster");
+             CertSigningRequest csr = OpenSSLUtil.createCsr(unsignedCluster);
+             CertPair cluster = OpenSSLUtil.signCrs(csr, Collections.singletonList("*.apps-crc.testing"), ca)
```
Comment (kornys) [inline systemtests/src/test/java/io/enmasse/systemtest/bases/web/ConsoleTest.java:1453]: this is really CRC specific, it cause issues when we will run it agains real cluster, better is use `Environment.getKubernetesDomain()`

File: systemtests/src/main/java/io/enmasse/systemtest/platform/Minikube.java [hunk 3]
```diff
@@ -15,12 +15,16 @@
 import io.fabric8.kubernetes.client.ConfigBuilder;
 import io.fabric8.kubernetes.client.DefaultKubernetesClient;
 import io.fabric8.kubernetes.client.utils.HttpClientUtils;
+import io.fabric8.openshift.api.model.Route;
```
Comment (kornys) [inline systemtests/src/main/java/io/enmasse/systemtest/platform/Minikube.java:18]: why we need imports in this class? I don't see any use in code

File: systemtests/src/test/java/io/enmasse/systemtest/bases/web/ConsoleTest.java (hunk 2 above)
Comment (k-wall) [reply systemtests/src/test/java/io/enmasse/systemtest/bases/web/ConsoleTest.java]: Fixed. I've now run the test on a proper OpenShift 4 cluster.  It passes.

File: systemtests/src/main/java/io/enmasse/systemtest/platform/Minikube.java (hunk 3 above)
Comment (k-wall) [reply systemtests/src/main/java/io/enmasse/systemtest/platform/Minikube.java]: fixed

**Comment by k-wall** [comment]
I've verified the changes on CRC, OpenShift 4 and checked for regressions on Minikube.

**Comment by k-wall** [comment]
@famartinrh @kornys re. CertificateUtils.  I didn't know about this class when I wrote the new test.  I will refactor the test classes to use a common openssl utility class, but don't intend to do this until after 0.31.1.   Okay?

**Comment by k-wall** [comment]
> > Should we support a way of explicitly configuring the console CA in the CRD as well? It could be useful both on Kubernetes and as an alternative fallback.
> 
> Yes. I think so. I'm planning to add providerTrustBundle that references a configmap. On OpenShift the controller will create the config, default its name and add the annotation (enmasse.io/ca-bundle-hash). On Kubernetes, the user can provide the configmap if he wishes, but it won't be mandatory. In the configmap, key(s that match pattern *.crt will be assumed to contain a PEM(s).
> 
> The configmap will be optionally mounted by console-init. It will aggregate all the trust into a single file and write to a location known to oauth-proxy. oauth-proxy will read use this file for its provider trust. There's a slight wrinkle with oauth-proxy - the OpenShift version requires a file, whereas the other one requires a cert directory.

@lulf  captured as a separate feature request. #4322

**Comment by k-wall** [comment]
@enmasse-ci run tests  testcase=FirefoxConsoleTest  ocp4

**Pull request merged**